pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,720 @@
+}
+
+// When a cluster-api cluster is created, enqueue the cluster deployment that owns it and update its expectations.
+func (c *Controller) addCluster(obj interface{}) {
+	cluster := obj.(*capi.Cluster)
+
+	if cluster.DeletionTimestamp != nil {
+		// on a restart of the controller manager, it's possible a cluster shows up in a state that
+		// is already pending deletion. Prevent the cluster from being a creation observation.
+		c.deleteCluster(cluster)
+		return
+	}
+
+	clusterDeployment, err := controller.ClusterDeploymentForCluster(cluster, c.clusterDeploymentsLister)
+	if err != nil {
+		clustoplog.WithCluster(c.logger, cluster).Errorf("error retrieving cluster deployment for cluster: %v", err)
+		return
+	}
+	if clusterDeployment == nil {
+		clustoplog.WithCluster(c.logger, cluster).Debugf("cluster is not controlled by a cluster deployment")
+		return
+	}
+	clustoplog.WithCluster(clustoplog.WithClusterDeployment(c.logger, clusterDeployment), cluster).Debugln("cluster created")
+	c.enqueueClusterDeployment(clusterDeployment)
+}
+
+// When a cluster is updated, figure out what cluster deployment manages it and wake it up.
+func (c *Controller) updateCluster(old, cur interface{}) {
+	oldCluster := old.(*capi.Cluster)
+	curCluster := cur.(*capi.Cluster)
+	if curCluster.ResourceVersion == oldCluster.ResourceVersion {
+		// Periodic resync will send update events for all known clusters.
+		// Two different versions of the same cluster will always have different RVs.
+		return
+	}
+
+	if curCluster.DeletionTimestamp != nil {
+		c.deleteCluster(curCluster)
+		return
+	}
+
+	clusterDeployment, err := controller.ClusterDeploymentForCluster(curCluster, c.clusterDeploymentsLister)
+	if err != nil {
+		clustoplog.WithCluster(c.logger, curCluster).Errorf("error retrieving cluster deployment for cluster: %v", err)
+		return
+	}
+	if clusterDeployment == nil {
+		clustoplog.WithCluster(c.logger, curCluster).Debugf("cluster is not controlled by a cluster deployment")
+		return
+	}
+	clustoplog.WithCluster(clustoplog.WithClusterDeployment(c.logger, clusterDeployment), curCluster).Debugln("cluster updated")
+	c.enqueueClusterDeployment(clusterDeployment)
+}
+
+// When a cluster is deleted, enqueue the cluster deployment that manages it and update its expectations.
+func (c *Controller) deleteCluster(obj interface{}) {
+	cluster, ok := obj.(*capi.Cluster)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Couldn't get object from tombstone %#v", obj))
+			return
+		}
+		cluster, ok = tombstone.Obj.(*capi.Cluster)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Tombstone contained object that is not a Cluster %#v", obj))
+			return
+		}
+	}
+
+	clusterDeployment, err := controller.ClusterDeploymentForCluster(cluster, c.clusterDeploymentsLister)
+	if err != nil {
+		clustoplog.WithCluster(c.logger, cluster).Errorf("error retrieving cluster deployment for cluster: %v", err)
```

Comment:
This might be what you referred to in your question about deletion but I also saw that this triggers on normal clusterdeployment delete.